pkg/controller/cloneset/sync/cloneset_scale.go

```diff
@@ -403,3 +420,101 @@ func (r *realControl) choosePodsToDelete(cs *appsv1alpha1.CloneSet, totalDiff in
 
 	return podsToDelete
 }
+
+func getInstanceIDsFromPods(pods []*v1.Pod) sets.String {
+	ins := sets.NewString()
+	for _, pod := range pods {
+		if id := clonesetutils.GetInstanceID(pod); id != "" {
+			ins.Insert(id)
+		}
+	}
+	return ins
+}
+
+func getUselessPVCs(pods []*v1.Pod, pvcs []*v1.PersistentVolumeClaim) map[string]*v1.PersistentVolumeClaim {
+	activeIDs := getInstanceIDsFromPods(pods)
+
+	useless := map[string]*v1.PersistentVolumeClaim{}
+	for _, pvc := range pvcs {
+		id := clonesetutils.GetInstanceID(pvc)
+		if id != "" && !activeIDs.Has(id) {
+			useless[id] = pvc
+		}
+	}
+	return useless
+}
+
+func (r *realControl) cleanupPVCs(cs *appsv1alpha1.CloneSet, uselessPVCs map[string]*v1.PersistentVolumeClaim) (bool, error) {
+	var modified bool
+
+	pods, err := getAllPods(r.Client, cs)
```

Comment:
why list again, can we use the pods param in Scale directly?

Reply:
Active pods are passed in from the Scale, so there is no way to get pods that are in the terminating state.

Reply:
it make senses now, yet since the Scale already pass pods variables, list pods again can break the code encapsulation logic, maybe we can move the pvc cleanup logic up onto doReconcile

Reply:
I have considered this. The reason for putting it in scale is because doReconcile seems to do some data preparation work. In fact, if “all pods” can be passed into Scale then there is no need to list pods, but the cost of changing the interface is a bit high.